### PR TITLE
DEPR: group by one element list gets scalar keys

### LINKED ISSUE #58858: BUG: Inconsistent types for groupby group names
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd

df = pd.DataFrame({'x': [10, 20, 30], 'y': ['a', 'b', 'c']})

# group by atomic - names and keys are atomic
name, _ = next(iter(df.groupby('x')))
name    # `10` as expected
df.groupby('x').groups  # `{10: [0], 20: [1], 30: [2]}` as expected

# group by list of 2 - names and keys are 2-tuples
name, _ = next(iter(df.groupby(['x', 'y'])))
name  # `(10, 'a')` as expected
df.groupby(['x', 'y']).groups  # `{(10, 'a'): [0], (20, 'b'): [1], (30, 'c'): [2]}` as expected

# group by list of 1 - names are 1-tuples but keys are atomic!?
name, _ = next(iter(df.groupby(['x'])))
name  # `(10,)` as expected
df.groupby(['x']).groups  # `{10: [0], 20: [1], 30: [2]}` UNEXPECTED
```

### Issue Description

In pandas 1.5.x grouping by a single element yielded scalar group names, grouping by a list of more than one element yielded tuples names but grouping by a length-1 list yielded scalar group names again which was considered inconsistent (and I agree).

This was mostly resolved in 2.0 (see #42795) and now when [iterating through groups](https://pandas.pydata.org/docs/user_guide/groupby.html#iterating-through-groups) the group names are 1-tuples when grouped with a length-1 list. However the [groups attribute](https://pandas.pydata.org/docs/user_guide/groupby.html#groupby-object-attributes) still returns scalars instead of 1-tuples which I consider to be a bug or at least requiring enhancement.  

### Expected Behavior

I expect group names / keys to be 1-tuples when grouping with a length-1 list: group by scalar get scalar names, group by list get tuple names. (I would also propose that grouping by an empty list should return a single group of the whole dataframe, with an empty tuple as the key, but thats out of scope for this issue!)

I expect the same group name / key behaviour whether  [iterating through groups](https://pandas.pydata.org/docs/user_guide/groupby.html#iterating-through-groups) or accessing groups via the [groups attribute](https://pandas.pydata.org/docs/user_guide/groupby.html#groupby-object-attributes). 

### Installed Versions

<details>
<summary>latest release (2.2.2)</summary>

INSTALLED VERSIONS
------------------
commit                : d9cdd2ee5a58015ef6f4d15c7226110c9aab8140
python                : 3.10.14.final.0
python-bits           : 64
OS                    : Linux
OS-release            : 5.15.0-1064-azure
Version               : #73~20.04.1-Ubuntu SMP Mon May 6 09:43:44 UTC 2024
machine               : x86_64
processor             : x86_64
byteorder             : little
LC_ALL                : None
LANG                  : C.UTF-8
LOCALE                : en_US.UTF-8

pandas                : 2.2.2
numpy                 : 1.26.4
pytz                  : 2024.1
dateutil              : 2.9.0.post0
setuptools            : 65.5.0
pip                   : 24.0
Cython                : None
pytest                : None
hypothesis            : None
sphinx                : None
blosc                 : None
feather               : None
xlsxwriter            : None
lxml.etree            : None
html5lib              : None
pymysql               : None
psycopg2              : None
jinja2                : None
IPython               : None
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : None
bottleneck            : None
dataframe-api-compat  : None
fastparquet           : None
fsspec                : None
gcsfs                 : None
matplotlib            : None
numba                 : None
numexpr               : None
odfpy                 : None
openpyxl              : None
pandas_gbq            : None
pyarrow               : None
pyreadstat            : None
python-calamine       : None
pyxlsb                : None
s3fs                  : None
scipy                 : None
sqlalchemy            : None
tables                : None
tabulate              : None
xarray                : None
xlrd                  : None
zstandard             : None
tzdata                : 2024.1
qtpy                  : None
pyqt5                 : None

</details>

<details>

<summary>dev install (3.0.0.dev0+1024.gb16233155)</summary>

INSTALLED VERSIONS
------------------
commit                : b162331554d7c7f6fd46ddde1ff3908f2dc8bcce
python                : 3.10.14.final.0
python-bits           : 64
OS                    : Linux
OS-release            : 5.15.0-1064-azure
Version               : #73~20.04.1-Ubuntu SMP Mon May 6 09:43:44 UTC 2024
machine               : x86_64
processor             : x86_64
byteorder             : little
LC_ALL                : None
LANG                  : C.UTF-8
LOCALE                : en_US.UTF-8

pandas                : 3.0.0.dev0+1024.gb16233155
numpy                 : 1.26.4
pytz                  : 2024.1
dateutil              : 2.9.0.post0
setuptools            : 65.5.0
pip                   : 24.0
Cython                : None
pytest                : None
hypothesis            : None
sphinx                : None
blosc                 : None
feather               : None
xlsxwriter            : None
lxml.etree            : None
html5lib              : None
pymysql               : None
psycopg2              : None
jinja2                : None
IPython               : None
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : None
bottleneck            : None
fastparquet           : None
fsspec                : None
gcsfs                 : None
matplotlib            : None
numba                 : None
numexpr               : None
odfpy                 : None
openpyxl              : None
pyarrow               : None
pyreadstat            : None
python-calamine       : None
pyxlsb                : None
s3fs                  : None
scipy                 : None
sqlalchemy            : None
tables                : None
tabulate              : None
xarray                : None
xlrd                  : None
zstandard             : None
tzdata                : 2024.1
qtpy                  : None
pyqt5                 : None

</details>

### PR DESCRIPTION
**Pull request by natmokval**
- [x] closes #58858

changed the behavior: group `keys` to be `1-tuples` instead of `scalar` when grouping with a `length-1 list`

### CODE REVIEW

**Comment by natmokval** [comment]
I deprecated the behavior for `groups`: group `keys `are now 1-tuples instead of scalars when grouping with a length-1 list.
I fixed the tests except for this one `test_groupby_raises_category_on_category` in `pandas/tests/groupby/test_raises.py`. @rhshadrach, could you please advise me on how to fix this test?

File: pandas/core/groupby/grouper.py [hunk 1]
```diff
@@ -667,6 +671,15 @@ def groups(self) -> dict[Hashable, Index]:
...
         cats = Categorical.from_codes(codes, uniques, validate=False)
+        if not self.key_dtype_str:
+            warnings.warn(
+                "`groups` by one element list returns scalar is deprecated "
+                "and will be removed. In a future version `groups` by one element "
+                "list will return tuple. Use ``df.groupby(by='a').groups`` "
+                "instead of ``df.groupby(by=['a']).groups`` to avoid this warning",
+                FutureWarning,
+                stacklevel=find_stack_level(),
+            )
```
Comment (rhshadrach) [inline pandas/core/groupby/grouper.py:682]: Can you move this to `GroupBy.groups`? Then you won't need to modify `Grouping` (e.g. include extra arguments) at all.

**Comment by rhshadrach** [comment]
> @rhshadrach, could you please advise me on how to fix this test?

Sorry for the delay! I believe you've resolved this, let me know if not.

**Comment by natmokval** [comment]
> Sorry for the delay! I believe you've resolved this, let me know if not.

Thanks @rhshadrach for reviewing this PR. I moved the FutureWarning from Grouping to GroupBy.groups as you suggested and added a note to v3.0.0. Could you please take a look at my changes?

File: pandas/tests/groupby/test_groupby.py [hunk 2]
```diff
@@ -2996,3 +2996,15 @@ def test_groupby_multi_index_codes():
 
     index = df_grouped.index
     tm.assert_index_equal(index, MultiIndex.from_frame(index.to_frame()))
+
+
+def test_groupby_keys_1length_list():
```
Comment (rhshadrach) [inline pandas/tests/groupby/test_groupby.py:3001]: Since this behavior is already hit in tests, I don't think we need to add a new test here.

**Review by rhshadrach — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/test_groupby.py (hunk 2 above)
Comment (natmokval) [reply pandas/tests/groupby/test_groupby.py]: thanks, I removed the new test

File: pandas/tests/groupby/test_grouping.py [hunk 3]
```diff
@@ -719,15 +732,18 @@ def test_grouping_labels(self, multiindex_dataframe_random_data):
...
         df.iloc[-1] = pd.NaT
         grouper = Grouper(key="date", freq="YS")
+        msg = "`groups` by one element list returns scalar is deprecated"
 
         # Grouper in a list grouping
         result = df.groupby([grouper])
         expected = {Timestamp("2011-01-01"): Index(list(range(364)))}
-        tm.assert_dict_equal(result.groups, expected)
+        with tm.assert_produces_warning(FutureWarning, match=msg):
+            result = result.groups
```
[the remark this answers is not in the document]
Comment (natmokval) [reply pandas/tests/groupby/test_grouping.py]: thanks, done

**Review by rhshadrach — APPROVED** [review]
lgtm

**Comment by rhshadrach** [comment]
Thanks @natmokval!

**Pull request merged**